Set scope in expand_abbrev rather than rely on update_scope

garrigue opened the pull request:
Having `update_scope` raise scopes recursively limits the usages of the scope field (see #13770, #13771).
However, the scope must be propagated to children when expanding GADT equations in order to check principality.

This PR modifies `Ctype.expand_abbrev_gen` so that it sets the scope of non-variable type nodes when expanding a type abbreviation, only in principal mode (but not limited to GADTs).
In turn, this requires adding a `?scope` argument to `Ctype.subst`, `Ctype.instance_parameterized_type` and `Ctype.copy` (where the real work is done).
I also switched to labelled arguments for the former (there are now 9 arguments).

The testsuite is left unchanged, and hopefully this should satsify @gasche 's need.

An open issue is whether to also apply this to non-principal mode.
The cost of setting the scope being low (it just requires re-expanding cached type abbreviations when the scope has changed, a supposedly rare situation), performance does not seem to be the problem.
This may break some (strange) existing programs, but is conservatism worth the discrepancy between the modes.

Two more points:
* There are some small concerns about the interaction with the abbreviation mechanism, which does not properly feedbacks scope updates on the expanded form to the original type, but this was already the case before; need to go back to keep_expansion for that.
* I also merged a comment into `with_local_level_generalize`, which may be of interest if someone plans to use the scope field for type variables. Namely, the scope of type variables is not checked when lowering their level in structure mode.

garrigue committed:
set scope in expand_abbrev rather than rely on update_scope

garrigue committed:
use Int.max rather than generic max

gasche commented:
Thanks! I find this indeed a nice direction to move in, and you can do it better than I would have -- I was planning to look at it but I'm happy to review rather than fumble my way around the codebase.

Naive question: you are tweaking the `subst` call in expand_abbrev_gen, but naively I would expect this to be equivalent to calling `subst` unchanged, and then calling an `update_scope_deep` function to fix the scopes afterwards, recursively. Wouldn't the latter approach also work and be sensibly less invasive to implement?

I also have a question on your new remark on the generalization step in `with_local_level_gen`: why not indeed check scopes at this point? I would naively expect it to be an invariant that scopes can never be strictly more restrictive than levels, with an error (immediate or delayed) when that happens. Is there a good reason not to fail with an escape in this case, besides the fact that this specific scenario cannot currently happen?

gasche commented:
Side note: my working idea to implement this broad feature was to update the scopes not during expansions, but at the point where we store the type expression in the type environment. We know that the type definition is never going to leave the generalization level of `equations_scope`, so it should be fine to lower it in the environment, and I would expect `subst`/copy-family functions to just copy scopes unchanged.

I don't have a particular preference between restricting scopes at environment-insertion time or at environment-lookup time, whatever is simpler and has your preference works.

gasche commented:
On the question of whether to do this in non-principal mode as well: if we are convinced that this is a good restriction to ambivalence, I would support enabling this in non-principal mode as well, even though it will break some code, to be fixed by adding more annotations. (I suspect that this pattern is more common than we think.)

If we want to allow more programs in non-principal mode, I think that my idea of specifying that we take the "oldest equations in scope" (that introduces the structure we need, here an arrow type) makes sense -- it is a deterministic policy/heuristic to choose among a finite set of options, that can be justified as the least-ambivalent choice. It may in fact be exactly the behavior that (non-principal) OCaml implements today, due to the way equations are inserted in-order in the typing environment.

(This is a roundabout way to say that both options are reasonable to me: the extra `-principal` restriction enforces that *any* use of an equation is tracked, while the current more permissive behavior has a specification that makes sense to me so it's not too bad.)

garrigue commented:
@gasche 
> Naive question: you are tweaking the `subst` call in expand_abbrev_gen, but naively I would expect this to be equivalent to calling `subst` unchanged, and then calling an `update_scope_deep` function to fix the scopes afterwards, recursively. Wouldn't the latter approach also work and be sensibly less invasive to implement?

Actually, this is not strictly equivalent, particularly if the type we are expanding has parameters.
The recursive `update_scope` approach raises all the scopes in the type, while the `subst` approach only raises the scopes of types in the body of the definition, and leaves the parameters unchanged. This is similar to the way recursive types are built from abbreviations, and it seems to make more sense. However, this has no direct impact for GADTs, as scoped equations  are closed anyway (they have no type parameters).
From an efficiency point of view, it also seems better to produce nodes with the right scope to start with than to update them later on.
However, I have no strong preference here.

> I also have a question on your new remark on the generalization step in `with_local_level_gen`: why not indeed check scopes at this point? I would naively expect it to be an invariant that scopes can never be strictly more restrictive than levels, with an error (immediate or delayed) when that happens. Is there a good reason not to fail with an escape in this case, besides the fact that this specific scenario cannot currently happen?

Not a very strong reason. It is just not very clear how this error should be processed: the code is in `Ctype`, but the error, if it happens, should probably be reported in `Typecore`.
If we add a mechanism that requires this check, we need to do something about it, i.e. probably change the API for `Ctype.with_local_level_generalize_structure`.

> Side note: my working idea to implement this broad feature was to update the scopes not during expansions, but at the point where we store the type expression in the type environment.

It do not think it would be sufficient. Namely, we do not only want the nodes to have a scope higher than the equation they are in, but also to have a scope higher than the type they have expanded, which is only known at expansion time.

gasche commented:
> Actually, this is not strictly equivalent, particularly if the type we are expanding has parameters.

Right, this is a good point, but I find the change a bit invasive. (A subset of the copy/substitution functions, those that are used during expansion, now carry a scope. If we change the functions we use there it's easy to forget to add a scope parameter to the new functions and introduce subtle bugs.)

> Namely, we do not only want the nodes to have a scope higher than the equation they are in, but also to have a scope higher than the type they have expanded, which is only known at expansion time.

I don't see why that is the case, do you have an example? My intuition is that propagating scopes in the (closed/rigid) result of an expansion is a way to mark all new type expressions appearing in the type as depending on this equation, so it makes sense to propagate only the scope of the equation. (Also: I must lack imagination but I cannot see how the immediate scope of an expandable type could be higher, more restrictive than the scope of an equation over it. Are you thinking of a case where we expand a parametrized type rather than just a variable?)

garrigue commented:
> > Actually, this is not strictly equivalent, particularly if the type we are expanding has parameters.
> 
> Right, this is a good point, but I find the change a bit invasive. (A subset of the copy/substitution functions, those that are used during expansion, now carry a scope. If we change the functions we use there it's easy to forget to add a scope parameter to the new functions and introduce subtle bugs.)

I do not see how it would happen. If a function takes a `~scope` argument, you are supposed to honor it, not just discard it. There is no problem in having functions that do not take a scope, as long as they are not on the path from `expand_abbrev` to `copy`. For me the main advantage of this approach is to make clear that we follow the same policy as `copy`. The other approach is to do something that is more specifically for GADTs (i.e. always raise the scopes of all nodes). This said, it may be better to hide the scope argument in the exported version of `instance_parameterized_type`, since it has probably no use outside of `Ctype`.

> > Namely, we do not only want the nodes to have a scope higher than the equation they are in, but also to have a scope higher than the type they have expanded, which is only known at expansion time.
> 
> I don't see why that is the case, do you have an example? My intuition is that propagating scopes in the (closed/rigid) result of an expansion is a way to mark all new type expressions appearing in the type as depending on this equation, so it makes sense to propagate only the scope of the equation. (Also: I must lack imagination but I cannot see how the immediate scope of an expandable type could be higher, more restrictive than the scope of an equation over it. Are you thinking of a case where we expand a parametrized type rather than just a variable?)

Suppose that you have two equations, phi1: `a = bool -> b` and phi2: `b = int -> int`, with the scope of phi1 being higher than that of phi2. If you need to expand `a` into `bool -> int -> int`, all nodes should have scope phi1. On the other hand if the scope of phi2 is higher, `bool` should have scope phi1, and `int` should have scope phi2.

gasche commented on typing/ctype.ml:
```diff
@@ -1620,7 +1627,8 @@ let expand_abbrev_gen kind find_type_expansion env ty =
         if level <> generic_level then update_level env level ty';
-        update_scope scope ty';
+        if not !Clflags.principal then update_scope scope ty'
+        else if get_scope ty' <> scope then raise_scope_escape_exn ty;
```
I would find it simpler to just call `update_scope` here unconditionally, as before; it should be equivalent, right?

garrigue replied:
Thinking of it, I'm not even sure it is correct to do that in the non-principal case.
Namely, cached abbreviations are shared between different parts of the type graph.
It might be that we expand several times the same path in places with a different scope.
In that case, this code forces the highest scope on all occurrences, which might be incorrect.
In the principal case, we rather re-expand the abbreviation, which has no such problem.
The only problem would be if it prevents closing a cycle because of different scopes inside an expansion, but I do not see how it could happen.
In the principal case, we could indeed also use the recursive version of `update_scope`, but it has the above potential problem, and does more raising than a re-expansion.

garrigue replied:
After a last look, always updating is probably fine, but it is only required if we actually use the type, which is not the case in principal mode anymore.
There might be some strange behavior due to the non-uniformity of the expansion; again I do not see how to fix that without keep-expansion.

gasche commented on typing/ctype.ml:
```diff
@@ -1636,19 +1644,23 @@ let expand_abbrev_gen kind find_type_expansion env ty =
+                 as this is not necessary for soundness *)
+              let scope = if !Clflags.principal then Some scope else None in
+              subst ~env ~level ?scope ~priv:kind ~abbrev ~oty:ty
+                ~params ~args body
```
I have the impression that the logic here is not quite right for my purposes of separating the GADT-specific scope handling and the general scope handling.

If I have an type definition in the context that does not come from a GADT, I do not expect its expansion to affect scopes at all. (In the implementation, its expansion scope is `Btype.lowest_scope`). With the current code as I understand it, the rigid part of the definition would still see its scope raised to the scope of the Tconstr type expression being expanded.

In particular, this may be incompatible with the approach I had in mind with inline record types, as the expansion of an inline record with a raised scope to a record definition would result in record field types at this higher scope -- as in my previous attempts.

Thinking about this more, I have the impression that we use scopes for two different mechanisms today:
- scopes for GADTs are intended by its designers to be "sticky", they track the fact that a type *somehow* depends on an equation; if we expand the type or take a sub-component or any other operation from it, the result should also inherit the sticky scope
- scopes for local type constructors are designed to be non-sticky, the constructor itself should not escape the current scope but it is valid to eliminate it by expansion -- so the expansion result should not have a raised scope.

If this analysis is correct, then it is actually important that we handle scopes differently for type constructors and for type expressions, and it may be (as you suggested earlier) a losing game to try to use type-expression scopes for inline record escapes, because (I think?) I want the non-sticky kind.

garrigue replied:
Indeed, this scope propagation is required because an equation may contain type abbreviations.

I agree that scopes are handled differently for type constructors and type expressions, but I do not see how this is a problem for inline record types. They are nominal types, so they will not be expanded, and as a result the scope will not be propagated to subexpressions. I don't remember all the details, but it may be that they are re-exported; in that case we would still be ok because this is an expansion from record type to record type, and the type parameters of inline records have no constraints, so they would be left alone.

gasche replied:
You are indeed correct, I checked by rebasing my PR on top of the present one and it works -- I pushed the result.

gasche reviewed (approved):
I have discussed this quite a bit with @garrigue and I am now convinced that this approach is reasonable: it preserves the current behavior, while getting rid of the systematic raising-scopes-in-depth behavior that blocks me for #13770.

I would be in favor of merging, if @garrigue also agrees.

garrigue commented:
I think that this approach is better than the original one, but I need to convince myself that everything is fine.
Give me just a bit more time before merging.

gasche commented:
@garrigue I did add new tests in https://github.com/ocaml/ocaml/pull/13770#issuecomment-2656820530 to check the cases where an inline record is re-exported from a previous type definition. Is this what you needed to feel better about your present PR?

garrigue committed:
add log and comments; small change in instance_parameterized_type

garrigue commented:
I have rebased and introduced a tiny change in `instance_parameterized_type`: rather than not updating the scope for type variables in `copy`, just exclude type parameters, as they are already part of the type before expansion.
@gasche You may want to check that this still works for your application. It should be fine since it reduces the updates.
Otherwise, this is ready to merge now.

gasche commented:
Thanks @garrigue! I decided that this was overall a good change, so I went ahead and merged in trunk.
